File: backend/models/Document.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

const Document = sequelize.define('Document', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  title: {
    type: DataTypes.STRING(500),
    allowNull: false,
    validate: {
      notEmpty: {
        msg: 'Название документа не может быть пустым'
      }
    }
  },
  category: {
    type: DataTypes.STRING(100),
    allowNull: false
  },
  year: {
    type: DataTypes.INTEGER,
    allowNull: false,
    validate: {
      min: {
        args: [1900],
        msg: 'Год должен быть больше 1900'
      },
      max: {
        args: [new Date().getFullYear() + 1],
        msg: 'Год не может быть в будущем'
      }
    }
  },
  pages: {
    type: DataTypes.INTEGER,
    allowNull: false,
    validate: {
      min: {
        args: [1],
        msg: 'Количество страниц должно быть больше 0'
      }
    }
  },
  price: {
    type: DataTypes.INTEGER,
    allowNull: false,
    comment: 'Цена в сумах',
    validate: {
      min: {
        args: [0],
        msg: 'Цена не может быть отрицательной'
      }
    }
  },
  description: {
    type: DataTypes.TEXT,
    allowNull: true
  },
  file_path: {
    type: DataTypes.STRING(500),
    allowNull: true,
    comment: 'Путь к полному PDF файлу'
  },
  preview_path: {
    type: DataTypes.STRING(500),
    allowNull: true,
    comment: 'Путь к preview PDF (первые 2 страницы)'
  },
  is_active: {
    type: DataTypes.BOOLEAN,
    defaultValue: true
  },
  download_count: {
    type: DataTypes.INTEGER,
    defaultValue: 0
  }
}, {
  tableName: 'documents',
  timestamps: true
});

module.exports = Document;

File: backend/app/models/document.py
from datetime import datetime, timezone

from flask import current_app

from app import db


class Document(db.Model):
    __tablename__ = "documents"

    id = db.Column(db.Integer, primary_key=True)
    title = db.Column(db.String(500), nullable=False)
    category = db.Column(db.String(100), nullable=False, index=True)
    year = db.Column(db.Integer, nullable=False)
    pages = db.Column(db.Integer, nullable=False)
    description = db.Column(db.Text)
    file_path = db.Column(db.String(500))  # path to PDF/Word on QNAP
    is_active = db.Column(db.Boolean, default=True)
    created_at = db.Column(
        db.DateTime, default=lambda: datetime.now(timezone.utc)
    )

    purchases = db.relationship("Purchase", backref="document", lazy="dynamic")

    @property
    def price(self):
        """Price = pages * PRICE_PER_PAGE (1000 sum per page)."""
        per_page = current_app.config.get("PRICE_PER_PAGE", 1000)
        return self.pages * per_page

    def to_dict(self):
        return {
            "id": self.id,
            "title": self.title,
            "category": self.category,
            "year": self.year,
            "pages": self.pages,
            "price": self.price,
            "description": self.description,
            "file_path": self.file_path,
            "is_active": self.is_active,
        }
